### PR TITLE
[MAINTENANCE] Resolve MetricResolutionErrors introduced by Expectation Pydantic annotations

### PR DESCRIPTION
**Pull request by cdkini**
Metrics ids are calculated based on the values noted in the Expectation config - note that `1.0 != 1` (these generate different values). Since Pydantic coerces types, we often saw integer inputs being converted to floats, creating mismatches when it came to grabbing metrics.

The changes here bring back the previous annotations we used (just `float` instead of `int | float`) but also resolve some other issues so we won't see this problems moving forward.

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

**Comment by ghost** [comment]
<details open>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=9064&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

####

[Review these changes using an interactive CodeSee Map](https://app.codesee.io/r/reviews?pr=9064&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
CodeSee Map legend
</details>

File: tests/checkpoint/conftest.py [hunk 1]
```diff
@@ -97,11 +98,12 @@ def titanic_pandas_data_context_stats_enabled_and_expectation_suite_with_one_exp
     context = titanic_pandas_data_context_with_v013_datasource_with_checkpoints_v1_with_empty_store_stats_enabled
     # create expectation suite
     suite = context.add_expectation_suite("my_expectation_suite")
-    expectation = ExpectationConfiguration(
-        expectation_type="expect_column_values_to_be_between",
-        kwargs={"column": "col1", "min_value": 1, "max_value": 2},
+    expectation = ExpectColumnValuesToBeBetween(
+        column="col1",
+        min_value=1,
+        max_value=2,
```
Comment (cdkini) [inline tests/checkpoint/conftest.py:104]: This was the culprit - we added an integer-based config during test setup but the Pydantic model coerces to floats.

This resulted in metric resolution errors because the generated ids in metrics code relies on the value (`1.0 != 1`)

A user will never touch configs so this is purely a test scaffolding issue.

**Review by tyler-hoffman — APPROVED** [review]
LGTM! Side note (not for this PR!): We should make a `Comparable` class/alias to encapsulate the list of what we support. Seems like we should have only had to update that in one place.

File: tests/checkpoint/test_checkpoint_result_format.py [hunk 2]
```diff
@@ -131,44 +144,28 @@ def reference_sql_checkpoint_config_for_multi_column_sum_table(
 
 
 @pytest.fixture()
-def expectation_config_expect_multicolumn_sum_to_equal() -> ExpectationConfiguration:
-    return ExpectationConfiguration(
-        expectation_type="expect_multicolumn_sum_to_equal",
-        kwargs={"column_list": ["a", "b", "c"], "sum_total": 30},
-    )
+def expect_multicolumn_sum_to_equal() -> Expectation:
+    return ExpectMulticolumnSumToEqual(column_list=["a", "b", "c"], sum_total=30)
```
Comment (cdkini) [inline tests/checkpoint/test_checkpoint_result_format.py:148]: Changed all of these fixtures from configs to rich types so we use the Pydantic coercion as intended

File: great_expectations/rule_based_profiler/expectation_configuration_builder/expectation_configuration_builder.py [hunk 3]
```diff
@@ -143,7 +160,7 @@ def _build_expectation_configuration(
         variables: Optional[ParameterContainer] = None,
         parameters: Optional[Dict[str, ParameterContainer]] = None,
         runtime_configuration: Optional[dict] = None,
-    ) -> ExpectationConfiguration:
+    ) -> ExpectationConfiguration | None:
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/expectation_configuration_builder/expectation_configuration_builder.py:163]: Type annotation was simply wrong before - child class has a nullable return

File: great_expectations/rule_based_profiler/rule/rule_output.py [hunk 4]
```diff
@@ -54,13 +54,15 @@ def get_expectation_configurations(self) -> List[ExpectationConfiguration]:
...
-                expectation_configurations.append(
+                config = (
                     expectation_configuration_builder.build_expectation_configuration(
                         domain=domain,
                         variables=self.rule_state.variables,
                         parameters=self.rule_state.parameters,
                     )
                 )
+                if config:
+                    expectation_configurations.append(config)
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule/rule_output.py:65]: Same note as above - we need to conditionally add based on null

File: great_expectations/rule_based_profiler/expectation_configuration_builder/expectation_configuration_builder.py [hunk 5]
```diff
@@ -107,9 +108,25 @@ def build_expectation_configuration(  # noqa: PLR0913
...
+    def _roundtrip_config_through_expectation(
+        self, domain: Domain, config: ExpectationConfiguration
+    ) -> ExpectationConfiguration:
+        """
+        Utilize Pydantic validaton and type coercion to ensure the final expectation configuration is valid.
+        """
+        expectation_cls = get_expectation_impl(config.expectation_type)
+        kwargs = {**config.kwargs, **domain.domain_kwargs}
+        expectation = expectation_cls(**kwargs, meta=config.meta)
+        return expectation.configuration
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/expectation_configuration_builder/expectation_configuration_builder.py:129]: Not my favorite solution but I don't fully understand how the expectation configuration builder creates config - let's simply roundtrip them through their expectation so we get proper validation and coercion

File: great_expectations/rule_based_profiler/expectation_configuration_builder/default_expectation_configuration_builder.py [hunk 6]
```diff
@@ -349,7 +349,7 @@ def _build_expectation_configuration(
         variables: Optional[ParameterContainer] = None,
         parameters: Optional[Dict[str, ParameterContainer]] = None,
         runtime_configuration: Optional[dict] = None,
-    ) -> Optional[ExpectationConfiguration]:
+    ) -> ExpectationConfiguration | None:
```
Comment (tyler-hoffman) [inline great_expectations/rule_based_profiler/expectation_configuration_builder/default_expectation_configuration_builder.py]: Just curious: was this change necessary, or just a style thing?

File: great_expectations/rule_based_profiler/expectation_configuration_builder/expectation_configuration_builder.py (hunk 5 above)
Comment (tyler-hoffman) [reply great_expectations/rule_based_profiler/expectation_configuration_builder/expectation_configuration_builder.py:129]: Yeah, agreed both that it's not the ideal, but that it's pragmatic.

File: great_expectations/rule_based_profiler/expectation_configuration_builder/default_expectation_configuration_builder.py (hunk 6 above)
Comment (cdkini) [reply great_expectations/rule_based_profiler/expectation_configuration_builder/default_expectation_configuration_builder.py]: just a style thing

File: tests/checkpoint/test_checkpoint_result_format.py (hunk 2 above)
Comment (tyler-hoffman) [reply tests/checkpoint/test_checkpoint_result_format.py:148]: Seems great. I mostly skimmed the rest of this file because it looks like a pretty standard find/replace.

**Review by tyler-hoffman — APPROVED** [review]

File: tests/rule_based_profiler/test_rule_output.py [hunk 7]
```diff
@@ -15,7 +15,7 @@ def test_rule_output_get_get_expectation_configurations(
     expected_expectation_configurations: List[ExpectationConfiguration] = [
         ExpectationConfiguration(
             **{
-                "expectation_type": "expect_my_validation",
+                "expectation_type": "expect_column_to_exist",
```
Comment (cdkini) [inline tests/rule_based_profiler/test_rule_output.py:18]: We need a real expectation type here now that we instantiate the expectation
